&duydl [MNT] Ensure Update Contributors does not run on main

Greyisheep opened the pull request:
#### Reference Issues/PRs
Fixes #6181.
<!--
Fixes #6181.

Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests.
If no issue exists, you can open one here: https://github.com/sktime/sktime/issues
-->


#### What does this implement/fix? Explain your changes.
This implement ensures that `Update Contributors` does not run on `main`. It also makes sure `build wheels` only triggers when `release` is triggered and and `Update Contributors` is completed.
<!--
A clear and concise description of what you have implemented.
-->

#### Did you add any tests for the change?
Not yet, but I am considering to add based on feedback and possible merge

<!-- This section is useful if you have added a test in addition to the existing ones. This will ensure that further changes to these files won't introduce the same kind of bug. It is considered good practice to add tests with newly added code to enforce the fact that the code actually works. This will reduce the chance of introducing logical bugs.
-->

#### Any other comments?
I think the workflow is too tightly woven together, and can be broken down further to ensure seamlessness.

Greyisheep commented:
@fkiraly 

Please, check linked PR, and give feedback

fkiraly reviewed (changes requested):
Thanks! Very useful!

Few questions/remarks:

* I think you are accidentally adding empty spaces, linting will fail
* can you explain what the addition in the release workflow `wheels.yml` will do?
* I'd rather not couple the "update contributors" workflow to the release workflow, imo when the release button is being pushed, at that point we should assume that the contributors list is up-to-date.

Greyisheep commented:
Thanks for the feedback.

**Addressing your remarks:**

> I think you are accidentally adding empty spaces, linting will fail
I will fix that right away

> can you explain what the addition in the release workflow `wheels.yml` will do?
The intention of the workflow there is to ensure that the "normal CI" is triggered if and only if a release is published (which was there previously), and the `Update Contributors` workflow is completed. My assumption of normal CI here is release and build wheels

Writing this now, I realized that what you referenced as "normal CI" in the issue, was a vaguely understood by me.

I then would like to ask, what particular CI workflow or action did you refer to as "normal CI" [here](https://github.com/sktime/sktime/issues/6181)

> I'd rather not couple the "update contributors" workflow to the release workflow, imo when the release button is being pushed, at that point we should assume that the contributors list is up-to-date.
`on` what would you like the update contributors to be triggered?
Not so sure what you mean here, but would removing the modification of wheels.yml and updating the name from `update_contributors.yml` to `00_update_contributors.yml`

fkiraly commented:
> My assumption of normal CI here is release and build wheels

No, `wheels.yml` is only triggered for release.

The CI that runs on pull requests is `test.yml`.

fkiraly commented:
Sure.

Currently, `update-contributors` and `test` lead to different workflow runs, e.g., here:
![image](https://github.com/sktime/sktime/assets/7985502/cfdc15d8-49ab-49e0-ae4d-ba4aa080e9ca)

It seems that the `test` workflow triggers first, so the GitHub UI wil only register the latest workflow run for checking whether tests have passed on the PR, to validate whether required jobs have passed.

That is, while things like `test-nosoftdeps` run on the `test` action, the outcome becomes irrelevant for the purpose of informing the Github UI, because it's looking for the job in the `update-contributors` workflow.

Consequently, the jobs dashboard on the pull request looks like this:
![image](https://github.com/sktime/sktime/assets/7985502/c4bf7645-9477-4ff9-9d85-2a314c4a59f1)
even if all the jobs are run (but in the second-to-last workflow call form the PR), and no merge button appears.

If the `update-contributors` workflow were to run first, and then `test`, this would not be a problem - but it is unclear to me how to ensure that.

sammychinedu2ky commented:
@fkiraly please where does it look for the job in the update-contributors workflow?

fkiraly commented:
> @fkiraly please where does it look for the job in the update-contributors workflow?

 @sammychinedu2ky, could you kindly be more specific? what looks for which job?

sammychinedu2ky commented:
> That is, while things like `test-nosoftdeps` run on the `test` action, the outcome becomes irrelevant for the purpose of informing the Github UI, because it's looking for the job in the `update-contributors` workflow.

You said this . so I am trying to understand what you meant by `looking for the job in the update-contributors` workflow

fkiraly commented:
Ah, I see. The "looking for" is not part of the workflow, but part of the GitHub UI's behaviour, i.e., which yellow dots are displayed to turn into green tickmarks etc, below.

sammychinedu2ky commented:
@fkiraly, why is it looking for the update-contributions workflow when it wasn't set to listen for a general push trigger? Also, I noticed that the flow requires approval from a maintainer to be executed. Maybe it will change to green after that.
![image](https://github.com/sktime/sktime/assets/36219292/786e92cf-ac92-4d92-b5ad-5ddced34c503)

fkiraly commented:
this is a different issue - CI on this PR does not rune because of approval.

The issue with `update-contributors` only occurs when it is triggered, namely once the `all-contributorsrc` is changed (which currently does not happen in this PR).

fkiraly commented:
The problem is, *if it runs*, then this happens:
https://github.com/sktime/sktime/pull/6189#issuecomment-2015095003

Let me know if you would like me to explain any part of the "what happens" in more detail.

sammychinedu2ky commented:
> Consequently, the jobs dashboard on the pull request looks like this: ![image](https://private-user-images.githubusercontent.com/7985502/316023790-c4bf7645-9477-4ff9-9d85-2a314c4a59f1.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTExMzM4NTAsIm5iZiI6MTcxMTEzMzU1MCwicGF0aCI6Ii83OTg1NTAyLzMxNjAyMzc5MC1jNGJmNzY0NS05NDc3LTRmZjktOWQ4NS0yYTMxNGM0YTU5ZjEucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDMyMiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAzMjJUMTg1MjMwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MmVhZTA3ZGU3OTg3N2IyYWMwMmRjYmFkYzZhZDRkNDZhMGUxNDdmM2M5ZGY2MWU3N2I5ODgyOWZjNzdiZDMyNSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.QtbT-0biVfd3vam5drO9clVN0BM1ZylZUo1c0nD8jGs) even if all the jobs are run (but in the second-to-last workflow call form the PR), and no merge button appears.
> 
> If the `update-contributors` workflow were to run first, and then `test`, this would not be a problem - but it is unclear to me how to ensure that.

@fkiraly  please can you explain this part

fkiraly commented:
A *workflow run* consists of multiple *job runs*, you can inspect these when clicking "details". The *job runs* are those displayed in the GitHub dashboard. *workflow runs* and *job runs* are instances of the general *workflow* and *job* (specification) in the `yml`.

Does this help?

sammychinedu2ky commented:
> A _workflow run_ consists of multiple _job runs_, you can inspect these when clicking "details". The _job runs_ are those displayed in the GitHub dashboard. _workflow runs_ and _job runs_ are instances of the general _workflow_ and _job_ (specification) in the `yml`.
> 
> Does this help?

Ooh I think I get it now. The problem is not related to this particular action for this pr. But the two workflow that will be triggered anytime a change is made to the contribution Json file right

sammychinedu2ky commented:
Will adding an exclusion path to the contribution file fix this? @fkiraly

fkiraly commented:
why would that help, what is your reasoning?

sammychinedu2ky commented:
It will prevent a push to the all-contributorsrc from triggering the test workflow. An listener for a pull request can also be added

fkiraly commented:
that wouldn't solve the issue, because we're looking at typical PR where both that file, and others are changed.

Quick question, @sammychinedu2ky, are you identical with @Greyisheep, or a different person?

sammychinedu2ky commented:
😂 no I'm not. We are different @fkiraly

fkiraly commented:
hm, ok, may I then suggest discussing further alternative approaches in the root issue #6181? Discussion in PR should be on the content of the PR, the code, primarily.

Greyisheep commented:
> Sure.
> 
> Currently, `update-contributors` and `test` lead to different workflow runs, e.g., here: ![image](https://private-user-images.githubusercontent.com/7985502/316023154-cfdc15d8-49ab-49e0-ae4d-ba4aa080e9ca.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTExNzA1MTUsIm5iZiI6MTcxMTE3MDIxNSwicGF0aCI6Ii83OTg1NTAyLzMxNjAyMzE1NC1jZmRjMTVkOC00OWFiLTQ5ZTAtYWU0ZC1iYTRhYTA4MGU5Y2EucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDMyMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAzMjNUMDUwMzM1WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YzIxOGRkNjIwNWVjYjEzYzU0YmEyZTVjNGYzODYyNmQxYjRhOTE0NjY2MDI1ZjlmZjA0YWUxZjM4MmU3ZGJmYSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.7Tjmm0qnKIEuU4ilH9szSxzccpu7IFffjeV_otKvs2o)
> 
> It seems that the `test` workflow triggers first, so the GitHub UI wil only register the latest workflow run for checking whether tests have passed on the PR, to validate whether required jobs have passed.
> 
> That is, while things like `test-nosoftdeps` run on the `test` action, the outcome becomes irrelevant for the purpose of informing the Github UI, because it's looking for the job in the `update-contributors` workflow.
> 
> Consequently, the jobs dashboard on the pull request looks like this: ![image](https://private-user-images.githubusercontent.com/7985502/316023790-c4bf7645-9477-4ff9-9d85-2a314c4a59f1.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTExNzA1MTUsIm5iZiI6MTcxMTE3MDIxNSwicGF0aCI6Ii83OTg1NTAyLzMxNjAyMzc5MC1jNGJmNzY0NS05NDc3LTRmZjktOWQ4NS0yYTMxNGM0YTU5ZjEucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDMyMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAzMjNUMDUwMzM1WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NTRhZDFiZjBkZmRiNTk5NGRiOWViZDFlOWViY2JlNWVjZDlkNzJlMjFiZmExMTA5NDY2YjIyYjRkZjllNmY2OCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.FAXahIfCyuSHxUkGiNfAVbwJgVnldYg3pFMa1VI6KkA) even if all the jobs are run (but in the second-to-last workflow call form the PR), and no merge button appears.
> 
> If the `update-contributors` workflow were to run first, and then `test`, this would not be a problem - but it is unclear to me how to ensure that.

@fkiraly I think I will partition this into separate PRs: One for making sure update contributors doesn't run on main, and the other problem which is this back and forth with @sammychinedu2ky  can be resolved separately, so we can all brainstorm together.

Greyisheep commented:
> that wouldn't solve the issue, because we're looking at typical PR where both that file, and others are changed.
> 
> Quick question, @sammychinedu2ky, are you identical with @Greyisheep, or a different person?

Different people

yarnabrina reviewed (approved):
Looks right.

It seems @duydl suggested this already, should we given them credit as well?

https://github.com/sktime/sktime/issues/6181#issuecomment-2012275116

fkiraly commented:
> It seems @duydl suggested this already, should we given them credit as well?

Yes, that was one day earlier, agree. Will add him as a second contributor in the change log.
